packages/flutter/lib/src/material/text_form_field.dart

```diff
@@ -72,4 +74,69 @@ class TextFormField extends FormField<String> {
       );
     },
   );
+
+  /// Controls the text being edited.
+  ///
+  /// If null, this widget will create its own [TextEditingController].
+  final TextEditingController controller;
+
+  @override
+  _TextFormFieldState createState() => new _TextFormFieldState();
+}
+
+class _TextFormFieldState extends FormFieldState<String> {
+  TextEditingController _controller;
+
+  TextEditingController get _effectiveController => widget.controller ?? _controller;
+
+  @override
+  TextFormField get widget => super.widget;
+
+  @override
+  void initState() {
+    super.initState();
+    if (widget.controller == null) {
+      _controller = new TextEditingController(text: widget.initialValue);
+    } else {
+      widget.controller.text = widget.initialValue;
+      widget.controller.addListener(_handleControllerChanged);
+    }
+  }
+
+  @override
+  void didUpdateWidget(TextFormField oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller == null && oldWidget.controller != null)
+      _controller = new TextEditingController.fromValue(oldWidget.controller.value);
+    else if (widget.controller != null && oldWidget.controller == null)
+      _controller = null;
+
+    if (widget.controller != oldWidget.controller) {
+      if (oldWidget.controller != null)
+        oldWidget.controller.removeListener(_handleControllerChanged);
+      if (widget.controller != null)
+        widget.controller.addListener(_handleControllerChanged);
+    }
```

Comment:
this function should probably be refactored so that you don't do redundant tests, and I think that will make my next comment moot, but:

Flutter style is to use `expr?.foo()` rather than `if (expr != null) expr.foo();`.

Reply:
Done.